DSCP option isn't actually marking packets

The DSCP marks functionality added in v1.55.0 doesn't seem to be working at all. @xenago and I tested on the following platform/version/backend combinations, and rclone didn't produce DSCP-marked packets on any of them:
* macOS 10.14 (Mojave), rclone 1.55.1, Google Drive IPv4 & IPv6
* Windows 10 20H2, rclone 1.55.1, Google Drive IPv4, Onedrive (personal) IPv4
* OpenSUSE Tumbleweed, rclone 1.55.0, Google Drive IPv4 & IPv6

I originally discovered the problem while using the restic integration, but the behaviour is the same on a simple copy too. I also used iperf3 to verify that DSCP tags weren't being mangled/washed by the network stack, and iperf3 emitted DSCP-tagged packets as expected.

**How to reproduce:**

1. Start tcpdump/Wireshark/tshark
2. Invoke rclone with the `--dscp LE` argument (works for all possible values, not just `LE`). For example: `rclone --dscp LE copy testfile remote:`. Optionally, add `--bind <address>` to force IPv4 on a dual-stack host.
3. After rclone finishes, check the DiffServ field of an outgoing packet to see that it isn't being marked (CS0). In Wireshark, this shows as `Differentiated Services Field: 0x00 (DSCP: CS0, ECN: Not-ECT)` for IPv4, and `Traffic Class: 0x00 (DSCP: CS0, ECN: Not-ECT)` for IPv6.

I've attached the following:
* A packet capture from the macOS example (IPv6) [rclone-dscp-bug-mac-v6.pcapng.gz](https://github.com/rclone/rclone/files/6402470/rclone-dscp-bug-mac-v6.pcapng.gz)
* The corresponding debug log [rclone-dscp-bug-mac.log](https://github.com/rclone/rclone/files/6402471/rclone-dscp-bug-mac.log)

Neither of us is familiar with golang, but I'll try to get a debug build running to narrow down the problem. I wanted to put it out there in case somebody more familiar with rclone/golang can spot the error faster.

**Edit:**

I put it into a debugger, and the problem appears to be in `dialer.go`: https://github.com/rclone/rclone/blob/9e2fbe0f1a9314e1330b4adc9d69de137be12ee6/fs/fshttp/dialer.go#L48-L68

I set breakpoints at lines 55-57. 55 and 56 are called as expected, but **line 57 is never reached**. Therefore, `SetTrafficClass`/`SetTOS` are never called.

<!--- Please keep the note below for others who read your bug report. -->

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

Ah-ha! If I reduce the if statement to just an assignment (`addr := c.RemoteAddr().(*net.IPAddr)`), I get this:

```
panic: interface conversion: net.Addr is *net.TCPAddr, not *net.IPAddr
```

So it seems like a type mismatch. Changing `c.RemoteAddr().(*net.IPAddr)` to `c.RemoteAddr().(*net.TCPAddr)` fixes the problem and emits DSCP-marked packets on IPv6 and IPv4 (tested on macOS). I can send a PR for this if it's the "right" solution after testing on Windows and Linux as well.

--- a/docs/content/docs.md
+++ b/docs/content/docs.md
@@ -787,6 +787,8 @@ rclone copy --dscp LE from:/from to:/to
 ```
 would make the priority lower than usual internet flows.
 
+This option has no effect on Windows (see [golang/go#42728](https://github.com/golang/go/issues/42728)).
+
 ### -n, --dry-run ###
 
 Do a trial run with no permanent changes.  Use this to see what rclone

--- a/fs/fshttp/dialer.go
+++ b/fs/fshttp/dialer.go
@@ -3,6 +3,7 @@ package fshttp
 import (
 	"context"
 	"net"
+	"strings"
 	"time"
 
 	"github.com/rclone/rclone/fs"
@@ -53,15 +54,14 @@ func (d *Dialer) DialContext(ctx context.Context, network, address string) (net.
 		return c, err
 	}
 	if d.tclass != 0 {
-		if addr, ok := c.RemoteAddr().(*net.IPAddr); ok {
-			if addr.IP.To16() != nil && addr.IP.To4() == nil {
-				err = ipv6.NewConn(c).SetTrafficClass(d.tclass)
-			} else {
-				err = ipv4.NewConn(c).SetTOS(d.tclass)
-			}
-			if err != nil {
-				return c, err
-			}
+		// IPv6 addresses must have two or more ":"
+		if strings.Count(c.RemoteAddr().String(), ":") > 1 {
+			err = ipv6.NewConn(c).SetTrafficClass(d.tclass)
+		} else {
+			err = ipv4.NewConn(c).SetTOS(d.tclass)
+		}
+		if err != nil {
+			return c, err
 		}
 	}
 	return newTimeoutConn(c, d.timeout)

--- a/fs/fshttp/dialer.go
+++ b/fs/fshttp/dialer.go
@@ -3,7 +3,9 @@ package fshttp
 import (
 	"context"
 	"net"
+	"runtime"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/rclone/rclone/fs"
@@ -46,6 +48,8 @@ func (d *Dialer) Dial(network, address string) (net.Conn, error) {
 	return d.DialContext(context.Background(), network, address)
 }
 
+var warnDSCPFail, warnDSCPWindows sync.Once
+
 // DialContext connects to the address on the named network using
 // the provided context.
 func (d *Dialer) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
@@ -59,9 +63,17 @@ func (d *Dialer) DialContext(ctx context.Context, network, address string) (net.
 			err = ipv6.NewConn(c).SetTrafficClass(d.tclass)
 		} else {
 			err = ipv4.NewConn(c).SetTOS(d.tclass)
+			// Warn of silent failure on Windows (IPv4 only, IPv6 caught by error handler)
+			if runtime.GOOS == "windows" {
+				warnDSCPWindows.Do(func() {
+					fs.LogLevelPrintf(fs.LogLevelWarning, nil, "dialer: setting DSCP on Windows/IPv4 fails silently; see https://github.com/golang/go/issues/42728")
+				})
+			}
 		}
 		if err != nil {
-			return c, err
+			warnDSCPFail.Do(func() {
+				fs.LogLevelPrintf(fs.LogLevelWarning, nil, "dialer: failed to set DSCP socket options: %v", err)
+			})
 		}
 	}
 	return newTimeoutConn(c, d.timeout)

